src/viam/sdk/tests/test_utils.hpp

```diff
@@ -12,5 +17,68 @@ using namespace viam::sdk;
 AttributeMap fake_map();
 std::vector<GeometryConfig> fake_geometries();
 
+// TestServer is a wrapper around viam::sdk::Server that is a friend of the
+// class and can therefore access its private fields.
+class TestServer {
+   public:
+    TestServer(std::shared_ptr<Server> sdk_server);
+    ~TestServer();
+
+    std::shared_ptr<grpc::Channel> grpc_in_process_channel(const grpc::ChannelArguments& args);
+
+   private:
+    std::shared_ptr<Server> sdk_server_;
+};
+
+// This function sets up the following architecture:
+// -- ResourceClient
+//        /\
+//
+//        | (grpc InProcessChannel)
+//
+//        \/
+// -- ResourceServer
+//        /\
+//
+//        | (function calls)
+//
+//        \/
+// -- MockResource
+//
+// This is as close to a real client->server->resource setup as we can get
+// without starting another process.
+//
+// The passed in test_case function will have access to the created ResourceClient.
+template <typename ClientType, typename F>
+void client_to_mock_pipeline(std::shared_ptr<Resource> mock, F&& test_case) {
+    // Create a ResourceManager. Add the mock resource to the ResourceManager.
+    // Create a Server. Use the mock's API to create a resource-specific
+    // server (like MotorServer) from the ResourceManager and Server. Start the
+    // Server.
+    auto rm = std::make_shared<ResourceManager>();
+    rm->add(mock->name(), mock);
+    auto server = std::make_shared<sdk::Server>();
+    auto rs = sdk::Registry::lookup_resource(mock->dynamic_api());
+    // resource_server is unused; we call create_resource_server to call
+    // register_service and associate the Server with the resource-specific
+    // server (like MotorServer).
+    auto resource_server = rs->create_resource_server(rm, *server);
+    server->start();
+
+    // Create a resource-specific client to the mock over an established
+    // in-process gRPC channel.
+    grpc::ChannelArguments args;
+    auto test_server = TestServer(server);
+    auto grpc_channel = test_server.grpc_in_process_channel(args);
+    ClientType resource_client(mock->name(), grpc_channel);
```

Comment:
I wanted to call `rs->create_rpc_client` here and avoid the need for `ClientType`, but `create_rpc_client` returns a `std::shared_ptr<Resource>` and I had trouble getting tests to use that as their "client".

Reply:
There may well be a clever way to avoid the need for `ClientType` but I don't know it, and I don't think it makes things particularly verbose such that we should stress about getting rid of it.